lib/processors/versionInfoGenerator.js

```diff
@@ -14,6 +16,423 @@ function getTimestamp() {
+		);
+		return false;
+	}
+	if ( !embeddedBy.length ) {
+		log.error(
+			"  component '%s': property 'sap.app/embeddedBy' has an empty string value (which is invalid), " +
+			"it won't be listed as 'embedded'", componentPath
+		);
+		return false;
+	}
+	let resolvedEmbeddedBy = posixPath.resolve(componentPath, embeddedBy);
+	if ( resolvedEmbeddedBy && !resolvedEmbeddedBy.endsWith("/") ) {
+		resolvedEmbeddedBy = resolvedEmbeddedBy + "/";
+	}
+	if ( libraryPathPrefix === resolvedEmbeddedBy ) {
+		log.verbose("  component's 'sap.app/embeddedBy' property points to library, list it as 'embedded'");
+		return true;
+	} else {
+		log.verbose(
+			"  component's 'sap.app/embeddedBy' points to '%s', don't list it as 'embedded'", resolvedEmbeddedBy
+		);
+		return false;
+	}
+};
+
+/**
+ * Retrieves the manifest path
+ *
+ * @param {string} filePath path to the manifest, e.g. "sap/x/manifest.json"
+ * @param {string} subPath relative sub path, e.g. "sdk"
+ * @returns {string} manifest path, e.g. "sap/x/sdk/manifest.json"
+ */
+const getManifestPath = (filePath, subPath) => {
+	const folderPathOfManifest = filePath.substr(0, filePath.length - "manifest.json".length) + subPath;
+	return posixPath.resolve(folderPathOfManifest + "/manifest.json");
+};
+
+/**
+ * Resolves the transitive dependencies recursively.
+ *
+ * @param {Map<string, DependencyInfo>} dependencyInfoMap
+ */
+const resolveTransitiveDependencies = (dependencyInfoMap) => {
+	dependencyInfoMap.forEach((libraryInfo) => {
+		libraryInfo.resolve(dependencyInfoMap);
+	});
+};
+
+class DependencyInfoObject {
+	/**
+	 *
+	 * @param {string} name name of the dependency, e.g. "sap.x"
+	 * @param {boolean} lazy lazy dependency
+	 */
+	constructor(name, lazy) {
+		this.name = name;
+		this.lazy = lazy;
+	}
+}
+
+class DependencyInfo {
+	/**
+	 *
+	 * @param {DependencyInfoObject[]} libs
+	 * @param {string} name
+	 */
+	constructor(libs, name) {
+		this.libs = libs;
+		this.name = name;
+
+		/**
+		 *
+		 * @type {DependencyInfoObject[]}
+		 */
+		this.libsResolved = [];
+		this.wasResolved = false;
+	}
+
+	isResolved(libName) {
+		return this.libsResolved.find((libResolved) => {
+			return libResolved.name === libName;
+		});
+	}
+
+	/**
+	 *
+	 * @param {string} libName
+	 * @param {boolean} lazy
+	 * @returns {DependencyInfoObject}
+	 */
+	addResolvedLibDependency(libName, lazy) {
+		let alreadyResolved = this.isResolved(libName);
+		if (!alreadyResolved) {
+			alreadyResolved = new DependencyInfoObject(libName, lazy);
+			this.libsResolved.push(alreadyResolved);
+		} else {
+			if (!alreadyResolved.lazy || !lazy) {
+				alreadyResolved.lazy = undefined;
+			}
+		}
+		return alreadyResolved;
+	}
+
+	/**
+	 *
+	 * @param {Map<string,DependencyInfo>} dependencyInfoMap
+	 * @param {boolean} [lazy] whether or not the dependency is lazy dependency which means
+	 *  all its dependencies should be treated as lazy
+	 */
+	resolve(dependencyInfoMap, lazy) {
+		if (!this.wasResolved || lazy) {
```

Comment:
The resolution is kept as state of this instance. It is only updated when lazy is true. 
But what when resolve is called again afterwards with lazy = false? Then the result will be the one from the last resolution with lazy = true - which is wrong.

Reply:
tackled with f14cba5